File: app/src/main/java/org/bspb/smartbirds/pro/ui/fragment/MonitoringListFragment.java
package org.bspb.smartbirds.pro.ui.fragment;

import android.app.ListFragment;
import android.app.LoaderManager;
import android.content.Context;
import android.content.CursorLoader;
import android.content.Loader;
import android.database.Cursor;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CursorAdapter;
import android.widget.ListView;

import org.androidannotations.annotations.EFragment;
import org.bspb.smartbirds.pro.R;
import org.bspb.smartbirds.pro.SmartBirdsApplication;
import org.bspb.smartbirds.pro.adapter.ModelCursorAdapter;
import org.bspb.smartbirds.pro.adapter.ModelCursorFactory;
import org.bspb.smartbirds.pro.content.Monitoring;
import org.bspb.smartbirds.pro.content.MonitoringManager;
import org.bspb.smartbirds.pro.db.MonitoringColumns;
import org.bspb.smartbirds.pro.ui.partial.MonitoringListRowPartialView;
import org.bspb.smartbirds.pro.ui.partial.MonitoringListRowPartialView_;

import static org.bspb.smartbirds.pro.db.SmartBirdsProvider.Monitorings;


/**
 * Created by groupsky on 08.03.17.
 */

@EFragment
public class MonitoringListFragment extends ListFragment implements LoaderManager.LoaderCallbacks<Cursor> {

    private static final String[] PROJECTION = {
            MonitoringColumns._ID,
            MonitoringColumns.CODE,
            MonitoringColumns.STATUS,
            MonitoringColumns.DATA,
    };
    private static final String TAG = SmartBirdsApplication.TAG + ".FMonLst";

    private CursorAdapter adapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "onCreate");
        super.onCreate(savedInstanceState);

        adapter = new ModelCursorAdapter<Monitoring>(getActivity(), R.layout.partial_monitoring_list_row, null, new ModelCursorFactory<Monitoring>() {
            @Override
            public Monitoring createModelFromCursor(Cursor cursor) {
                return MonitoringManager.monitoringFromCursor(cursor);
            }
        }) {
            @Override
            public View newView(Context context, Cursor cursor, ViewGroup parent) {
                return MonitoringListRowPartialView_.build(context);
            }

            @Override
            public void bindView(View view, Context context, Monitoring model) {
                if (!(view instanceof MonitoringListRowPartialView))
                    throw new IllegalArgumentException("Must use " + MonitoringListRowPartialView.class.getSimpleName());
                ((MonitoringListRowPartialView) view).bind(model);
            }
        };
        setListAdapter(adapter);
        getLoaderManager().restartLoader(0, null, this);
    }

    @Override
    public void onResume() {
        Log.d(TAG, "onResume");
        super.onResume();

        getLoaderManager().restartLoader(0, null, this);
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        super.onListItemClick(l, v, position, id);
        Monitoring monitoring = (Monitoring) getListAdapter().getItem(position);
        if (monitoring != null && getActivity() instanceof Listener) {
            ((Listener)getActivity()).onMonitoringSelected(monitoring.code);
        }
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        Log.d(TAG, "onCreateLoader");
        return new CursorLoader(getActivity(),
                Monitorings.CONTENT_URI,
                PROJECTION,
                MonitoringColumns.STATUS + "=?",
                new String[]{Monitoring.Status.finished.name()},
                MonitoringColumns._ID + " DESC");
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor cursor) {
        Log.d(TAG, String.format("onLoadFinished: %d", cursor.getCount()));
        adapter.swapCursor(cursor);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
        Log.d(TAG, "onLoaderReset");
        adapter.swapCursor(null);
    }

    public interface Listener {
        void onMonitoringSelected(String monitoringCode);
    }
}

File: app/src/main/kotlin/org/bspb/smartbirds/pro/ui/fragment/MonitoringListFragment.kt
package org.bspb.smartbirds.pro.ui.fragment

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.ListView
import androidx.fragment.app.ListFragment
import androidx.fragment.app.viewModels
import org.bspb.smartbirds.pro.R
import org.bspb.smartbirds.pro.SmartBirdsApplication
import org.bspb.smartbirds.pro.adapter.MonitoringListAdapter
import org.bspb.smartbirds.pro.content.Monitoring
import org.bspb.smartbirds.pro.viewmodel.MonitoringListViewModel

class MonitoringListFragment : ListFragment() {

    companion object {
        private const val TAG = SmartBirdsApplication.TAG + ".FMonLst"
        private const val KEY_STATUS = "status"
    }

    private var filterStatus: Monitoring.Status? = null

    private lateinit var menuStatusFinished: MenuItem
    private lateinit var menuStatusUploaded: MenuItem
    private lateinit var menuStatusCanceled: MenuItem
    private lateinit var menuStatusPaused: MenuItem

    private var adapter: MonitoringListAdapter? = null
    private val viewModel: MonitoringListViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(TAG, "onCreate")
        super.onCreate(savedInstanceState)
        savedInstanceState?.run {
            getString(KEY_STATUS)?.apply {
                filterStatus = Monitoring.Status.valueOf(this)
            }
        }

        adapter = MonitoringListAdapter(requireContext())
        listAdapter = adapter
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViewModel()
        setupListView()
    }

    private fun initViewModel() {
        viewModel.init()
        viewModel.entries?.observe(viewLifecycleOwner) { items ->
            items?.apply {
                adapter?.clear()
                adapter?.addAll(items)
                adapter?.notifyDataSetChanged()
            }
        }
    }

    override fun onResume() {
        Log.d(TAG, "onResume")
        super.onResume()
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        super.onListItemClick(l, v, position, id)
        val monitoring = listAdapter?.getItem(position) as Monitoring
        if (monitoring != null && activity is Listener) {
            (activity as Listener).onMonitoringSelected(monitoring.code)
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.monitoring_list, menu)
        menuStatusFinished = menu.findItem(R.id.menu_filter_status_finished)
        menuStatusUploaded = menu.findItem(R.id.menu_filter_status_uploaded)
        menuStatusCanceled = menu.findItem(R.id.menu_filter_status_canceled)
        menuStatusPaused = menu.findItem(R.id.menu_filter_status_paused)
        filterStatus?.apply {
            updateMenuItems(this)
        }
        super.onCreateOptionsMenu(menu, inflater)
    }

    @Deprecated("Deprecated in Java")
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        if (itemId == R.id.menu_filter_status_finished) {
            filterFinished()
            return true
        }
        if (itemId == R.id.menu_filter_status_uploaded) {
            filterUploaded()
            return true
        }
        if (itemId == R.id.menu_filter_status_paused) {
            filterPaused()
            return true
        }
        if (itemId == R.id.menu_filter_status_canceled) {
            filterCanceled()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    interface Listener {
        fun onMonitoringSelected(monitoringCode: String?)
    }

    private fun setupListView() {
        setEmptyText(getText(R.string.emptyList))
    }

    private fun filterFinished() {
        filterByStatus(Monitoring.Status.finished)
    }

    private fun filterUploaded() {
        filterByStatus(Monitoring.Status.uploaded)
    }

    private fun filterPaused() {
        filterByStatus(Monitoring.Status.paused)
    }

    private fun filterCanceled() {
        filterByStatus(Monitoring.Status.canceled)
    }

    private fun filterByStatus(status: Monitoring.Status) {
        filterStatus = if (filterStatus == status) {
            null
        } else {
            status
        }

        updateMenuItems(filterStatus)
        viewModel.setFilterStatus(filterStatus)
        activity?.title = if (filterStatus != null) {
            getString(R.string.title_monitoring_list) + " [" + getString(filterStatus!!.label) + "]"
        } else {
            getString(R.string.title_monitoring_list)
        }

    }

    private fun updateMenuItems(status: Monitoring.Status?) {
        menuStatusFinished.isChecked = false
        menuStatusUploaded.isChecked = false
        menuStatusPaused.isChecked = false
        menuStatusCanceled.isChecked = false

        status?.apply {
            when (this) {
                Monitoring.Status.finished ->
                    menuStatusFinished.isChecked = true

                Monitoring.Status.uploaded ->
                    menuStatusUploaded.isChecked = true

                Monitoring.Status.paused ->
                    menuStatusPaused.isChecked = true

                Monitoring.Status.canceled ->
                    menuStatusCanceled.isChecked = true

                else -> {}
            }
        }

    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putString(KEY_STATUS, filterStatus?.name)
    }
}